pontoon/base/models.py

```diff
@@ -2913,30 +2933,94 @@ def calculate_stats(self, save=True):
         # Singular
         translations = Translation.objects.filter(
             entity__in=translated_entities.filter(string_plural=''),
-            locale=locale
+            locale=locale,
+        ).prefetch_related(
+            'errors',
+            'warnings',
         )
-        approved = translations.filter(approved=True).count()
-        fuzzy = translations.filter(fuzzy=True).count()
-        missing = resource.total_strings - approved - fuzzy
+
+        approved = translations.filter(
+            approved=True,
+            errors__isnull=True,
+            warnings__isnull=True,
+        ).count()
+
+        fuzzy = translations.filter(
+            fuzzy=True,
+            errors__isnull=True,
+            warnings__isnull=True,
+        ).count()
+
+        errors = translations.filter(
+            Q(
+                Q(Q(approved=True) | Q(fuzzy=True)) &
+                Q(errors__isnull=False)
+            ),
+        ).count()
+
+        warnings = translations.filter(
+            Q(
+                Q(Q(approved=True) | Q(fuzzy=True)) &
+                Q(warnings__isnull=False)
+            ),
+        ).count()
+
         unreviewed = translations.filter(
             approved=False,
             fuzzy=False,
-            rejected=False
+            rejected=False,
         ).count()
 
+        missing = resource.total_strings - approved - fuzzy - errors - warnings
+
         # Plural
         nplurals = locale.nplurals or 1
         for e in translated_entities.exclude(string_plural=''):
-            translations = Translation.objects.filter(entity=e, locale=locale)
-            plural_approved_count = translations.filter(approved=True).count()
-            plural_fuzzy_count = translations.filter(fuzzy=True).count()
+            translations = Translation.objects.filter(
+                entity=e,
+                locale=locale,
+            ).prefetch_related(
```

Comment:
[nit/questions] I wonder if it's required do prefetch fields if you only call `count()` later.

Reply:
Good call!